### ISSUE TITLE
SASToken Expired Exception was not getting thrown by the Device Client Open or Send

### ISSUE THREAD

**Issue by jkriba** [Azure/azure-iot-sdk-java#94]
Hi Support,
We are using Android Version 7.1 for developing a Cloud connector for our business application in the mobile device. Unlike the Azure IoT  Device Client provided by the .Net SDK , in case of  using an expired SAS token to Open / Send  events to IoTHub the device client of Java SDK is not throwing any exceptions and still trying to send the message which is eventually resulting in UN Authorization of the packet.

- **OS and version used:** <VERSION> Android 7.1

- **Java runtime used:** <VERSION> JDK "1.8.0_45"

- **SDK version used:** <VERSION> IoT Device Client 1.1.26 for Java

# Description of the issue:
Azure IoT DeviceClient for Java is not throwing an exception when it uses an Expired SAS token to send an event to the Azure IoTHub. This was working perfect with the Microsoft .Net SDK i.e. DeviceClient . We were not able to manage the state of the application SAS Token without this exception handling.

# Code sample exhibiting the issue:
void CreateDeviceClient() throws TokenExpiredException, Exception
{
   try
   {
      _logger.LogDebug(GetConnectionString());

      //Connection string with expired SAS Token       
      deviceClient = new DeviceClient(GetConnectionString(),_protocol); 

     _logger.LogDebug("CreateDeviceClient() - created ");
   }
   catch (Exception e)
   {
      //Here the exception is not get caught like the .Net SDK for an Expired SAS Token
      throw e;
   }
}

# Console log of the issue:
D/AndroidLog: 05/10/2017 11:55:37 AM :: IoTHubTransportEx.java:CreateDeviceClient:223 :: CreateDeviceClient() - created 
D/AndroidLog: 05/10/2017 11:55:37 AM :: IoTHubTransportEx.java:OpenDeviceClientAsync:258 :: OpenDeviceClientAsync() - opened 
V/AndroidLog: 05/10/2017 11:55:37 AM :: ConnectionMonitor.java:SendHeartBeat:311 :: Enter
I/AndroidLog: 05/10/2017 11:55:37 AM :: ConnectionMonitor.java:SendHeartBeat:314 :: SendHeartbeatAsync() Sending heartbeat
V/AndroidLog: 05/10/2017 11:55:37 AM :: IotHubConnectionPlugin.java:Send:188 :: Enter
D/AndroidLog: 05/10/2017 11:55:37 AM :: IotHubConnectionPlugin.java:InternalSendAsync:228 :: D/AndroidLog: 05/10/2017 11:55:38 AM :: IoTHubTransportEx.java:execute:272 :: IoT Hub responded to message : HeartbeatRequest : 7f38d512-7918-4268-bfe6-07b13257843d with status UNAUTHORIZED

**Comment by prmathur-microsoft** [comment]
Thanks for logging this issue. We will work on fix for this.

**Comment by jkriba** [comment]
Thanks prmathur for the rapid response , looking forward towards the fix. 

Regards
JK

**Comment by timtay-microsoft** [comment]
This functionality is fixed now as of commit adacfce

Feel free to re-open if you run into any issues tied to this

**Comment by AnnapoorniS** [comment]
Will this fix give any Security Exception on SASToken Expiry for DeviceClient sendEventAsync() method?

**Comment by prmathur-microsoft** [comment]
Not currently. Currently we only throw for open when a SasToken is provided. In case of sendEventAsync, token is renewed if expired.

**Comment by jkriba** [comment]
Hi prmathur,
  Thanks for the fix, but it will not be helpful unless the functionality is working along with SendEventAsyc [Similar to .NET SDK] rather than Open or CreateDeviceClient since technically the open / CreateDeviceClient will be called only once. 
 Hence we would like to re open this issue since we need the functionality with SendEventAsyc as we are requesting for a very basic need here..This is impacting the reliability of the Microsoft Device SDK for Java , hope you can understand our Genuine concern,

Regards
JK

**Comment by prmathur-microsoft** [comment]
If a SasToken is expired during SendEventAsyc , it is renewed during the process and SendEventAsyc  then continues to send over the new connection established with renewed sastoken. So technically there would not be a time when this exception gets thrown from SendEventAsyc .Also you can control your sas token expiry time using set option provided by SDK. So can you explain the exact scenario where you might need this exception to be thrown?

**Comment by AnnapoorniS** [comment]
I/System.out: {"deviceId":"sample","messageId":0,"temperature":29.871072181646547,"humidity"35.9518364760222}
I/System.out: IoT Hub responded to message 0 with status UNAUTHORIZED
I/System.out: {"deviceId":"sample","messageId":1,"temperature":21.206727175615423,"humidity"39.006367381809184}
I/System.out: IoT Hub responded to message 1 with status UNAUTHORIZED
I/System.out: {"deviceId":"sample","messageId":2,"temperature":20.70356410338387,"humidity"40.76523431622604}
I/System.out: IoT Hub responded to message 2 with status UNAUTHORIZED
I/System.out: {"deviceId":"sample","messageId":3,"temperature":25.83686198183099,"humidity"45.87321863295474}
I/System.out: IoT Hub responded to message 3 with status UNAUTHORIZED
I/System.out: {"deviceId":"sample","messageId":4,"temperature":21.843089593269315,"humidity"49.46218349362101}
I/System.out: IoT Hub responded to message 4 with status UNAUTHORIZED

I tried to send data from an expired SAS token, it simply rejects the message without any exception being thrown from the device client.  There should be an exception thrown from the device client SendEventAsync() which can notify the user to renew the SAS token.

Thanks,
Annapoorni S

**Comment by prmathur-microsoft** [comment]
Did you connect with expire token and then sent data over it ? or Did the token expire after you connected and you suddenly started receiving UNAUTHORIZED ?

**Comment by AnnapoorniS** [comment]
I tried with the latest java device client SDK 1.3.29 for the scenario **token expiration after connected**(i.e.after opening the device client).
The device client sample was sending packets without any trouble until the SAS token expires and responses back to EventCallback with the response OK_EMPTY. But, suddenly after the time SAS token expires, there is no response back from the device client and the data is being sent continuously, this is where it is worse without any exception being thrown by the sendEventAsync() method.

I/System.out: {"deviceId":"sample","messageId":427,"temperature":25.17581322043474,"humidity"31.33070145513511}
I/System.out: IoT Hub responded to message 427 with status OK_EMPTY
I/System.out: {"deviceId":"sample","messageId":428,"temperature":23.548720127449876,"humidity"32.2215514268882}
I/System.out: IoT Hub responded to message 428 with status OK_EMPTY
I/System.out: {"deviceId":"sample","messageId":429,"temperature":22.122520924236653,"humidity"49.2370461302427}
I/System.out: IoT Hub responded to message 429 with status OK_EMPTY
I/System.out: {"deviceId":"sample","messageId":430,"temperature":29.014385472463935,"humidity"40.8771477805924}
I/System.out: IoT Hub responded to message 430 with status OK_EMPTY
I/System.out: {"deviceId":"sample","messageId":431,"temperature":21.02669126733285,"humidity"47.54509540270777}
I/System.out: {"deviceId":"sample","messageId":432,"temperature":24.495326904204273,"humidity"34.5208662827706}
I/System.out: {"deviceId":"sample","messageId":433,"temperature":27.40687016293576,"humidity"32.42690434205118}
I/System.out: {"deviceId":"sample","messageId":434,"temperature":24.649301293511073,"humidity"43.7973590600514}

**Comment by prmathur-microsoft** [comment]
If the above scenario your are talking about is for MQTT then this is a known issue on our end and we are working on fixing it.

**Comment by AnnapoorniS** [comment]
I sent the message using HTTPS protocol with java SDK 1.3.29.

**Comment by prmathur-microsoft** [comment]
We are investigating this issue and we will update you soon.

**Comment by prmathur-microsoft** [comment]
A fix for this issue has been recently checked into master branch at @35647317a30ee686a43a64bbb4b4f5ae5eb05c3a please give it a try and let us know if it fixes your issue. Maven packages will be released soon as well.

**Comment by prmathur-microsoft** [comment]
Maven package with the fix for this issue has been released. Closing this now. feel free to reopen if you see any further issues.

**Comment by sruthi06** [comment]
While validating the fix for SASTokenExpiration using the latest version of DeviceClient (1.4.34) using the microsoft sample, it works fine with HTTPS but gives un expected behavior or false results for AMQPS_WS.

1. While using HTTPS, when the SAS Token gets expired in between, it gives us UNAUTHORIZED as the status in callback for every message sent.
2. In case of AMQPS_WS , it was giving OK_EMPTY for the messages sent after the token got expired which is absolutely wrong / false behaviour and does not match the behavior of HTTPS
	 
	AMQPS_WS is the most used transport protocol for the production environment..

As summary , the fix was not consistent and reliable to move forward. Hence re open this particular issue...

**Comment by jkriba** [comment]
Hi,
  This behaviour has been even observed with AMQPS and MQTT too..
Hence fix is not reliable and consistent. 

Please re open the same, thanks

Regards
JK

**Comment by prmathur-microsoft** [comment]
You are right. This fix was only related to HTTP and we are yet to fix it for other protocols. Feel free to open another issue for rest of the protocols.

**Comment by timtay-microsoft** [comment]
I am working on getting this bug fixed across all protocols. ETA is that it should be released 10/20. I'll keep this thread updated as new information comes along.

**Comment by timtay-microsoft** [comment]
Sorry for the delay.

The fix for AMQPS/AMQPS_WS/MQTT/MQT_WS is checked in. You can expect it to be in the next release we do on 10/20.

You will now receive UNAUTHORIZED callbacks when the token has expired after the device client has been opened.

**Comment by timtay-microsoft** [comment]
Just a small update: The next release will be happening 10/23, not 10/20. Sorry for the delay.

**Comment by timtay-microsoft** [comment]
The fix for this issue is now released. See the release below for versions:

https://github.com/Azure/azure-iot-sdk-java/releases/tag/2017-10-23 

I will close this issue for now. Feel free to re-open if it did not fix your issues.

**Comment by sruthi06** [comment]
Hi,
	 We have validdated the SASTokenExpiration using the latest version of DeviceClinet (1.5.37) using the microsoft sample. While using HTTPS protocol, it gives an UNAUTHORIZED status as callback only once for every message sent after the token is expired. But incase of using AMQPS_WS, it infinitely gives UNAUTHORIZED as a callback for every message that is sent after the token gets expired.
AMQPS_WS is the most used transport protocol for the production environment..

Request you to kindly validate it.

[Logs - SAS Token Expiration.txt](https://github.com/Azure/azure-iot-sdk-java/files/1445188/Logs.-.SAS.Token.Expiration.txt)

Thanks,
Sruthi

**Comment by timtay-microsoft** [comment]
I see the issue. I'll try to get this fix out as soon as I can. Thanks for bringing it to our attention.

**Comment by timtay-microsoft** [comment]
The fix for this bug is checked in as of 5d8ad37784de9b795648608bd61aeeda018a8215

It will be in the next release

**Comment by timtay-microsoft** [comment]
@sruthi06 The fix for that bug has been released:

https://github.com/Azure/azure-iot-sdk-java/releases/tag/2017-12-2

I'm closing this issue for now, feel free to re-open if it something else comes up. Sorry for the delay.

**Comment by Eqbalazar** [comment]
I am using "com.microsoft.azure.sdk.iot:iot-device-client:1.9.0"  , When I run Microsoft Sample 
I am getting UNAUTHORIZED Error for each send message . 
See Error Log : 

03-02 11:50:15.041 4821-4821/com.iothub.azure.microsoft.com.androidsample I/System.out: {"deviceId":"MyJ3Device","messageId":0,"temperature":25.709301894759466,"humidity":31.291324764047108}
03-02 11:50:15.101 4821-5115/com.iothub.azure.microsoft.com.androidsample I/System.out: (HTTPLog)-Static: isSBSettingEnabled false
03-02 11:50:15.102 4821-5115/com.iothub.azure.microsoft.com.androidsample I/System.out: (HTTPLog)-Static: isSBSettingEnabled false
03-02 11:50:15.576 4821-5115/com.iothub.azure.microsoft.com.androidsample I/System.out: IoT Hub responded to message 0 with status UNAUTHORIZED
03-02 11:50:17.050 4821-4821/com.iothub.azure.microsoft.com.androidsample I/System.out: {"deviceId":"MyJ3Device","messageId":1,"temperature":26.101337504470717,"humidity":45.12018306648988}
03-02 11:50:17.059 4821-5115/com.iothub.azure.microsoft.com.androidsample I/System.out: (HTTPLog)-Static: isSBSettingEnabled false
03-02 11:50:17.060 4821-5115/com.iothub.azure.microsoft.com.androidsample I/System.out: (HTTPLog)-Static: isSBSettingEnabled false
03-02 11:50:17.344 4821-5115/com.iothub.azure.microsoft.com.androidsample I/System.out: IoT Hub responded to message 1 with status UNAUTHORIZED
03-02 11:50:19.054 4821-4821/com.iothub.azure.microsoft.com.androidsample I/System.out: {"deviceId":"MyJ3Device","messageId":2,"temperature":28.325664820110916,"humidity":30.2472952585952}
03-02 11:50:19.067 4821-5116/com.iothub.azure.microsoft.com.androidsample I/System.out: (HTTPLog)-Static: isSBSettingEnabled false
03-02 11:50:19.067 4821-5116/com.iothub.azure.microsoft.com.androidsample I/System.out: (HTTPLog)-Static: isSBSettingEnabled false
03-02 11:50:19.447 4821-5116/com.iothub.azure.microsoft.com.androidsample I/System.out: IoT Hub responded to message 2 with status UNAUTHORIZED
03-02 11:50:21.059 4821-4821/com.iothub.azure.microsoft.com.androidsample I/System.out: {"deviceId":"MyJ3Device","messageId":3,"temperature":22.864126928624902,"humidity":39.93069710287709}
03-02 11:50:21.075 4821-5115/com.iothub.azure.microsoft.com.androidsample I/System.out: (HTTPLog)-Static: isSBSettingEnabled false
03-02 11:50:21.075 4821-5115/com.iothub.azure.microsoft.com.androidsample I/System.out: (HTTPLog)-Static: isSBSettingEnabled false
03-02 11:50:21.603 4821-5115/com.iothub.azure.microsoft.com.androidsample I/System.out: IoT Hub responded to message 3 with status UNAUTHORIZED
03-02 11:50:23.065 4821-4821/com.iothub.azure.microsoft.com.androidsample I/System.out: {"deviceId":"MyJ3Device","messageId":4,"temperature":22.262881135794064,"humidity":48.595869688160576}
03-02 11:50:23.074 4821-5115/com.iothub.azure.microsoft.com.androidsample I/System.out: (HTTPLog)-Static: isSBSettingEnabled false
03-02 11:50:23.075 4821-5115/com.iothub.azure.microsoft.com.androidsample I/System.out: (HTTPLog)-Static: isSBSettingEnabled false
03-02 11:50:23.422 4821-5115/com.iothub.azure.microsoft.com.androidsample I/System.out: IoT Hub responded to message 4 with status UNAUTHORIZED

**Comment by timtay-microsoft** [comment]
@Eqbalazar That error message means that your SAS token has expired, or your x509 certificate is invalid. Could you share more details on your authentication mechanism?

**Comment by Eqbalazar** [comment]
@timtay-microsoft 
I changed ConnectionString and DeviceID in sample source code in order to connect my iotHUB. Also I changed IotHubClientProtocol to HTTPS ;
That's it . Should I change any where ?

**Comment by timtay-microsoft** [comment]
That is all you should need to change. What is the format of your connection string?

It should look like one of the following:
```HostName=<YOUR HUB NAME>.azure-devices.net;DeviceId=<YOUR DEVICE ID>;SharedAccessKey=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX=```

or

```HostName=<YOUR HUB NAME>.azure-devices.net;DeviceId=<YOUR DEVICE ID>;SharedAccessSignature=SharedAccessSignature sr=XXXXXXXXXXX&sig=XXXXXXXXXXXX&se=XXXXXXXXX```

If it is the second, that format is a SAS token and those do expire over time. Ensure that you are using a new SAS token

**Comment by timtay-microsoft** [comment]
That isn't a valid connection string format. Check the Azure portal for your device, then grab the device key connection string from the device details

**Comment by Eqbalazar** [comment]
@timtay-microsoft Thank you so much , Now I am able to send message .
